builtin/audit/file/backend.go

```diff
@@ -137,100 +144,118 @@ type Backend struct {
 	mode     os.FileMode
 
 	saltMutex  sync.RWMutex
-	salt       *salt.Salt
+	salt       *atomic.Value
 	saltConfig *salt.Config
 	saltView   logical.Storage
 }
 
 var _ audit.Backend = (*Backend)(nil)
 
 func (b *Backend) Salt(ctx context.Context) (*salt.Salt, error) {
-	b.saltMutex.RLock()
-	if b.salt != nil {
-		defer b.saltMutex.RUnlock()
-		return b.salt, nil
+	s := b.salt.Load().(*salt.Salt)
+	if s != nil {
+		return s, nil
 	}
-	b.saltMutex.RUnlock()
+
 	b.saltMutex.Lock()
 	defer b.saltMutex.Unlock()
-	if b.salt != nil {
-		return b.salt, nil
+
+	s = b.salt.Load().(*salt.Salt)
+	if s != nil {
+		return s, nil
 	}
-	salt, err := salt.NewSalt(ctx, b.saltView, b.saltConfig)
+
+	newSalt, err := salt.NewSalt(ctx, b.saltView, b.saltConfig)
 	if err != nil {
+		b.salt.Store((*salt.Salt)(nil))
 		return nil, err
 	}
-	b.salt = salt
-	return salt, nil
+
+	b.salt.Store(newSalt)
+	return newSalt, nil
 }
 
 func (b *Backend) GetHash(ctx context.Context, data string) (string, error) {
 	salt, err := b.Salt(ctx)
 	if err != nil {
 		return "", err
 	}
+
 	return audit.HashString(salt, data), nil
 }
 
 func (b *Backend) LogRequest(ctx context.Context, in *logical.LogInput) error {
-	b.fileLock.Lock()
-	defer b.fileLock.Unlock()
-
+	var writer io.Writer
 	switch b.path {
 	case "stdout":
-		return b.formatter.FormatRequest(ctx, os.Stdout, b.formatConfig, in)
+		writer = os.Stdout
 	case "discard":
-		return b.formatter.FormatRequest(ctx, ioutil.Discard, b.formatConfig, in)
+		return nil
 	}
 
-	if err := b.open(); err != nil {
+	buf := bytes.NewBuffer(make([]byte, 0, 2000))
+	err := b.formatter.FormatRequest(ctx, buf, b.formatConfig, in)
+	if err != nil {
 		return err
 	}
 
-	if err := b.formatter.FormatRequest(ctx, b.f, b.formatConfig, in); err == nil {
+	return b.log(ctx, buf, writer)
+}
+
+func (b *Backend) log(ctx context.Context, buf *bytes.Buffer, writer io.Writer) error {
+	reader := bytes.NewReader(buf.Bytes())
+
+	b.fileLock.Lock()
+
+	if writer == nil {
+		if err := b.open(); err != nil {
+			b.fileLock.Unlock()
+			return err
+		}
+		writer = b.f
+	}
+
+	if _, err := reader.WriteTo(writer); err == nil {
+		b.fileLock.Unlock()
 		return nil
+	} else if b.path == "stdout" {
+		b.fileLock.Unlock()
+		return err
 	}
 
-	// Opportunistically try to re-open the FD, once per call
+	// If writing to stdout there's no real reason to think anything would have
```

Comment:
Do we want to log the error that we are ignoring here, as a warning in server logs?

Reply:
Nvm

Reply:
Sure, we could.

Reply:
Actually -- we can't. We don't have a logger at this point. We could maybe add that in centrally at some point in the future.